symforce/codegen/backends/python/templates/util/util.jinja

```diff
@@ -217,3 +217,33 @@ def {{ function_name_and_args(spec) }}:
         {%- if not loop.last %}, {% endif %}
     {%- endfor -%}
 {% endmacro %}
+
+{# ------------------------------------------------------------------------- #}
+
+{# Macro to flatten an array if it's an ndarray and is a vector. Also, raises
+ # a ValueError if the length is not equal to size and the shape is not that of
+ # a vector.
+ #
+ # Args:
+ #     array (str): The name of an sequence that might be an ndarray we wish to flatten
+ #     size (int): The expected size of the array
+ #}
+{% macro flatten_if_ndarray(array, size) %}
+{% set size_is_not_prime = size not in (1, 2, 3, 5, 7, 11, 13, 17, 19, 23, 29) %}
+if isinstance({{ array }}, numpy.ndarray):
+    {% if size_is_not_prime %}
+    if not (
+        {{ array }}.shape == ({{ size }},)
+        or {{ array }}.shape == ({{ size }}, 1)
+        or {{ array }}.shape == (1, {{ size }})
+    ):
+        raise ValueError("Expected {{ array }} to be a vector of length {{ size }}; instead had shape {}".format({{ array }}.shape))
+    {% endif %}
+    {{ array }} = {{ array }}.flatten()
+{% if size_is_not_prime %}elif{% else %}if{% endif %} len({{ array }}) != {{ size }}:
```

Comment:
For my own sanity:

For prime sizes, this reduces to:
```python
if isinstance({{ array }}, numpy.ndarray):
    {{ array }} = {{ array }}.flatten()
if len({{ array }}) != {{ size }}:
    raise ValueError(
        "Expected {{ array }} to be a sequence of length {{ size }}, was instead length {}.".format(
            len({{ array }})
        )
    )
```

For "maybe prime or composite" sizes this reduces to:
```python
if isinstance({{ array }}, numpy.ndarray):
    if not (
        {{ array }}.shape == ({{ size }},)
        or {{ array }}.shape == ({{ size }}, 1)
        or {{ array }}.shape == (1, {{ size }})
    ):
        raise ValueError("Expected {{ array }} to be a vector of length {{ size }}; instead had shape {}".format({{ array }}.shape))
    {{ array }} = {{ array }}.flatten()
elif len({{ array }}) != {{ size }}:
    raise ValueError(
        "Expected {{ array }} to be a sequence of length {{ size }}, was instead length {}.".format(
            len({{ array }})
        )
    )
```

My issues are that for prime sizes

1) you won't get error messages telling you the shape of the thing you passed, which can be useful (and may not be obvious from the size of the thing, e.g. if you passed something (3,7) to a function that wanted a vector of size 3)

2) we'll accept arrays with more dimensions.  "This function accepts matrices of shape (N,), (N, 1), or (1, N), unless N is a prime number less than 31, in which case it accepts matrices of shape (1, ..., 1, N, 1, ..., 1)" seems like an unfortunate API

Reply:
Sorry for making you sort through that. I was more or less aware of these issues (well, I hadn't considered point 1), but hadn't yet digested the situation enough to decide what the best option was.

Anyway, I'm just treating everything as a composite now.